### PR TITLE
MapChunk Operand Closure and Callable cleanup

### PR DESCRIPTION
**Pull request by vcfgv**
## What do these changes do?

This PR extends application of Apply Operand Closure cleanup to DataFrameMapChunk Operand as well as Callable and furthermore calculates func size with more accuracy.

Lifecycle management of UDF was implemented in commit 243965c but will be reverted due to refinements introduced in PR #3254.

## Related issue number

#3193 

## Check code requirements

- [x] tests added / passed (if needed)
- [x] Ensure all linting tests pass, see [here](https://docs.pymars.org/en/latest/development/contributing.html#check-code-styles) for how to run them

### CODE REVIEW

File: mars/dataframe/utils.py [hunk 1]
```diff
@@ -1456,23 +1455,91 @@ def clean_up_func(op):
...
+    from collections import deque
+
+    BYPASS_CLASSES = (str, bytes, Number, range, bytearray, pd.DataFrame, pd.Series)
+
+    def check_exceed_threshold():
+        nonlocal threshold, counted_bytes
+        if counted_bytes >= threshold:
+            raise StopIteration
+
+    def getsize(obj_outer):
```
Comment (fyrestone) [inline mars/dataframe/utils.py:1482]: There are many sizeof utils:
- The `calc_data_size` in `mars/utils.py`
- The `asizeof` in asv.

What's the difference of this `getsize`?

File: mars/dataframe/utils.py [hunk 2]
```diff
@@ -1456,23 +1455,91 @@ def clean_up_func(op):
...
+
+    from numbers import Number
+    from collections import deque
+
+    BYPASS_CLASSES = (str, bytes, Number, range, bytearray, pd.DataFrame, pd.Series)
+
+    def check_exceed_threshold():
+        nonlocal threshold, counted_bytes
+        if counted_bytes >= threshold:
+            raise StopIteration
```
Comment (fyrestone) [inline mars/dataframe/utils.py:1480]: The `StopIteration` has special meanings. If it is raised inside an iteration loop, then the loop will be stopped. Use a `RuntimeError` or just the `Exception`.

File: mars/services/task/supervisor/preprocessor.py [hunk 3]
```diff
@@ -203,6 +205,11 @@ def tile(self, tileable_graph: TileableGraph) -> Iterable[ChunkGraph]:
                     optimize_chunk_graph(chunk_graph)
                 )
             yield chunk_graph
+        # Extract ids of tileables whose op need func cleanup.
+        for t in tileable_graph:
+            if hasattr(t.op, "func_key") and t.op.func_key is not None:
```
Comment (fyrestone) [inline mars/services/task/supervisor/preprocessor.py:210]: Do we need to check all the tileables again? You have checked the tileables a few lines above.

File: mars/services/task/supervisor/preprocessor.py [hunk 4]
```diff
@@ -203,6 +205,11 @@ def tile(self, tileable_graph: TileableGraph) -> Iterable[ChunkGraph]:
...
+                assert t.op.need_clean_up_func is True
+                self.tileable_ids_need_clean_up_func.append(t.key)
```
Comment (fyrestone) [inline mars/services/task/supervisor/preprocessor.py:212]: Is the `func_key` changed during execution? If not, append the `func_key` to `self.tileable_ids_need_clean_up_func` is better.

File: mars/dataframe/tests/test_utils.py [hunk 5]
```diff
@@ -644,3 +645,45 @@ def get_chunks_meta(self, data_keys: List[str], **_) -> List[Dict]:
     assert isinstance(s2.chunks[1].op, DataFrameConcat)
     assert s2.chunks[1].name == "a"
     assert len(s2.chunks[1].op.inputs) == 2
+
+
+@pytest.mark.parametrize("multiplier_and_expected", [(1, False), (3, True), (4, True)])
+def test_whether_to_clean_up(multiplier_and_expected):
```
Comment (chaokunyang) [inline mars/dataframe/tests/test_utils.py:651]:
```suggestion
def test_whether_to_clean_up(multiplier, expected):
```

File: mars/dataframe/utils.py (hunk 1 above)
Comment (vcfgv) [reply mars/dataframe/utils.py:1482]: This `getsize` takes more classes into account and is able to early stop.

File: mars/services/task/supervisor/processor.py [hunk 6]
```diff
@@ -456,3 +458,21 @@ def _finish(self):
...
+            self.tileable_id_to_tileable[tileable_id]
+            for tileable_id in self._preprocessor.tileable_ids_need_clean_up_func
+        ]
+        func_keys = [t.op.func_key for t in tileables_need_clean_up_func]
+        logger.debug(f"Remove func storage {func_keys}")
+        ctx = get_context()
+        # Note: Considering that remove_func_storage is executed while TaskProcessor is being
+        # destroyed, creating storage_api may fail. Error type that may occur is still under
+        # investigation as well as experiment.
+        storage_api = await StorageAPI.create(ctx.session_id, ctx.supervisor_address)
```
Comment (chaokunyang) [inline mars/services/task/supervisor/processor.py:473]: `function` are put into storage by `ctx.storage_put`, why not cleanup func by `ctx.storage_delete`

File: mars/services/task/supervisor/preprocessor.py (hunk 3 above)
Comment (chaokunyang) [inline mars/services/task/supervisor/preprocessor.py:210]:
```suggestion
            if getattr(t.op, "func_key", None) is not None:
```

File: mars/services/task/supervisor/processor.py [hunk 7]
```diff
@@ -456,3 +458,21 @@ def _finish(self):
 
     def is_done(self) -> bool:
         return self.done.is_set()
+
+    async def remove_func_storage(self):
+        tileables_need_clean_up_func = [
+            self.tileable_id_to_tileable[tileable_id]
+            for tileable_id in self._preprocessor.tileable_ids_need_clean_up_func
```
Comment (chaokunyang) [inline mars/services/task/supervisor/processor.py:465]: After `func` cleaned up, does `tileable_ids_need_clean_up_func` need to be clear too?

**Review by chaokunyang — APPROVED** [review]
LGTM

**Review by fyrestone — APPROVED** [review]
LGTM

**Pull request merged**